sos/plugins/juju_machine.py

```diff
@@ -0,0 +1,78 @@
+# This file is part of the sos project: https://github.com/sosreport/sos
+# Copyright (C) 2019 Nick Niehoff <nick.niehoff@canonical.com>
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import os
+from sos.plugins import Plugin, UbuntuPlugin
+
+
+class JujuMachine(Plugin, UbuntuPlugin):
+    """ Juju orchestration tool - Machine
+    """
+
+    plugin_name = "juju_machine"
+    profiles = ("virt", "sysmgmt")
+
+    # Using files instead of packages here because there is no identifying
+    # package on a juju machine
+    files = ("/var/log/juju", "/usr/bin/juju-run")
+
+    def setup(self):
+        # Get agent configs for each agent
+        agent_configs = []
+        for dirpath, dirnames, filenames in os.walk("/var/lib/juju/agents"):
+            for dirname in dirnames:
+                agent_configs.append(
+                    "/var/lib/juju/agents/" + dirname + "/agent.conf")
+        self.add_copy_spec(agent_configs)
+
+        # Get a directory listing of /var/log/juju and /var/lib/juju
+        self.add_cmd_output(
+            ["ls -alRh /var/log/juju*", "ls -alRh /var/lib/juju*"]
+        )
+
+        if self.get_option("all_logs"):
+            # /var/lib/juju used to be in the default capture moving here
+            # because it usually was way to big.  However, in most cases you
+            # want all logs you want this too.
+            self.add_copy_spec(["/var/log/juju", "/var/lib/juju"])
+        else:
+            # We need this because we want to collect to the limit of all
+            # logs in the directory.
+            self.add_copy_spec("/var/log/juju/*.log")
+
+    def postproc(self):
+        agents_path = "/var/lib/juju/agents/*"
+        protect_keys = [
+            "sharedsecret",
+            "apipassword",
+            "oldpassword",
+            "statepassword",
+        ]
+
+        protect_certs = [
+            "systemidentity",
+            "caprivatekey",
+            "controllerkey",
+            "controllercert",
+            "cacert",
+        ]
+
+        # Replace simple yaml style "key: value"
+        keys_regex = r"((?m)^\s*(%s)\s*:\s*)(.*)" % "|".join(protect_keys)
+        # Will match and replace a yaml multiline string that is a certificate
+        certs_regex = (
+            r"((?m)^\s*(%s)\s*:\s*)(\|\s*\n\s+-+BEGIN (.*)"
+            r"-+\s(\s+\S+\n)+\s+-+END )\4(-+)" % "|".join(protect_certs)
```

Comment:
We have `self.do_cmd_private_sub` which clears certs from command output. We should probably expand that to do file substitution as well, since I imagine we're doing something similar to this in other plugins as well and it'd be good to standardize all of this.

Reply:
I understand, but I think this falls out of the scope for this Pull request.

Reply:
Can't be `do_file_private_sub` used here? See https://github.com/sosreport/sos/commit/df6deb2ff7acd84afe2e5f2823565329af333018.

Reply:
The current `do_file_private_sub()` looks for fixed `"------BEGIN"` / `"------END"` blocks - from the regex used here it looks like these have another component to the string. We can extend the method to handle this case, but we'll need to see some examples of the blocks that need to be replaced in these files.